### PR TITLE
[DPE-2684] Juju3 pipelines

### PR DESCRIPTION
**Pull request by juditnovak**
## Issue

Current pipelines are only supporting Juju2

## Solution

Extending pipelines with a matrix also executing Juju3.
Note that this triggers usage of Juju Secrtets (internal so far), which requires changes on tests

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -75,7 +85,29 @@ jobs:
...
         groups: ${{ fromJSON(needs.gh-hosted-collect-integration-tests.outputs.groups) }}
-    name: (GH hosted) ${{ matrix.groups.job_name }}
+        juju-snap-channel:  ["2.9/stable", "3.1/candidate"]
+        include:
+          - juju-snap-channel: "3.1/candidate"
+            agent-version: "3.1.6"
+            libjuju-version:  "3.2.0.1"
+          - juju-snap-channel: "2.9/stable"
+            agent-version: "2.9.44.0"
+            libjuju-version:  "2.9.44.0"
```
Comment (paulomach) [inline .github/workflows/ci.yaml:95]: please bump to `2.9.44.1` as `.0` will break with python 3.10 (default on jammy)

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -94,6 +126,12 @@ jobs:
         uses: charmed-kubernetes/actions-operator@main
         with:
           provider: microk8s
+          channel: "1.28-strict/stable"
+          bootstrap-options: "--agent-version ${{ matrix.agent-version }}"
+          juju-channel: ${{ matrix.juju-snap-channel }}
+      - name: Update python-libjuju version
+        if: ${{ matrix.libjuju-version != '2.9.44.0' }}
```
Comment (paulomach) [inline .github/workflows/ci.yaml:133]: same as above
```suggestion
        if: ${{ matrix.libjuju-version != '2.9.44.1' }}
```

File: pyproject.toml [hunk 3]
```diff
@@ -57,7 +58,9 @@ pytest-github-secrets = {git = "https://github.com/canonical/data-platform-workf
 pytest-operator = "^0.28.0"
 pytest-operator-cache = {git = "https://github.com/canonical/data-platform-workflows", tag = "v5.0.0", subdirectory = "python/pytest_plugins/pytest_operator_cache"}
 pytest-operator-groups = {git = "https://github.com/canonical/data-platform-workflows", tag = "v5.0.0", subdirectory = "python/pytest_plugins/pytest_operator_groups"}
-juju = "^2.9.44.1"
+juju = "^2.9.44.0"
```
Comment (paulomach) [inline pyproject.toml:61]:
```suggestion
juju = "^2.9.44.1"
```

**Review by delgod — APPROVED** [review]

File: .github/workflows/ci.yaml [hunk 4]
```diff
@@ -75,7 +85,29 @@ jobs:
       fail-fast: false
       matrix:
         groups: ${{ fromJSON(needs.gh-hosted-collect-integration-tests.outputs.groups) }}
-    name: (GH hosted) ${{ matrix.groups.job_name }}
+        juju-snap-channel:  ["2.9/stable", "3.1/candidate"]
+        include:
+          - juju-snap-channel: "3.1/candidate"
+            agent-version: "3.1.6"
+            libjuju-version:  "3.2.0.1"
```
Comment (delgod) [inline .github/workflows/ci.yaml:92]:
```suggestion
            libjuju-version:  "3.2.2"
```

**Review by paulomach — APPROVED** [review]

File: .github/workflows/ci.yaml [hunk 5]
```diff
@@ -94,6 +126,12 @@ jobs:
         uses: charmed-kubernetes/actions-operator@main
         with:
           provider: microk8s
+          channel: "1.28-strict/stable"
```
[the remark this answers is not in the document]
Comment (paulomach) [reply .github/workflows/ci.yaml:129]: yes, no problem

**Review by delgod — APPROVED** [review]

**Pull request merged**